node-hub/dora-openai-websocket/src/main.rs

```diff
@@ -0,0 +1,474 @@
+
+    for chunk in bytes.chunks_exact(2) {
+        let pcm16_sample = i16::from_le_bytes([chunk[0], chunk[1]]);
+        let f32_sample = pcm16_sample as f32 / 32767.0;
+        samples.push(f32_sample);
+    }
+
+    samples
+}
+
+fn convert_f32_to_pcm16(samples: &[f32]) -> Vec<u8> {
+    let mut pcm16_bytes = Vec::with_capacity(samples.len() * 2);
+
+    for &sample in samples {
+        // Clamp to [-1.0, 1.0] and convert to i16
+        let clamped = sample.max(-1.0).min(1.0);
+        let pcm16_sample = (clamped * 32767.0) as i16;
+        pcm16_bytes.extend_from_slice(&pcm16_sample.to_le_bytes());
+    }
+
+    pcm16_bytes
+}
+
+/// Replaces a placeholder in a file and writes the result to an output file.
+///
+/// # Arguments
+///
+/// * `input_path` - Path to the input file with placeholder text.
+/// * `placeholder` - The placeholder text to search for (e.g., "{{PLACEHOLDER}}").
+/// * `replacement` - The text to replace the placeholder with.
+/// * `output_path` - Path to write the modified content.
+fn replace_placeholder_in_file(
+    input_path: &str,
+    placeholder: &str,
+    replacement: &str,
+    output_path: &str,
+) -> io::Result<()> {
+    // Read the file content into a string
+    let content = fs::read_to_string(input_path)?;
+
+    // Replace the placeholder
+    let modified_content = content.replace(placeholder, replacement);
+
+    // Write the modified content to the output file
+    let mut file = fs::File::create(output_path)?;
+    file.write_all(modified_content.as_bytes())?;
+
+    Ok(())
+}
+
+async fn handle_client(fut: upgrade::UpgradeFut) -> Result<(), WebSocketError> {
+    let mut ws = fastwebsockets::FragmentCollector::new(fut.await?);
+
+    let frame = ws.read_frame().await?;
+    if frame.opcode != OpCode::Text {
+        return Err(WebSocketError::InvalidConnectionHeader);
+    }
+    let data: OpenAIRealtimeMessage = serde_json::from_slice(&frame.payload).unwrap();
+    let OpenAIRealtimeMessage::SessionUpdate { session } = data else {
+        return Err(WebSocketError::InvalidConnectionHeader);
+    };
+
+    let input_audio_transcription = session
+        .input_audio_transcription
+        .map_or("moyoyo-whisper".to_string(), |t| t.model);
+    let id = random::<u16>();
+    let node_id = format!("server-{id}");
+    let dataflow = format!("{input_audio_transcription}-{}.yml", id);
+    let template = format!("{input_audio_transcription}-template-metal.yml");
+    println!("Filling template: {}", template);
+    replace_placeholder_in_file(&template, "NODE_ID", &node_id, &dataflow).unwrap();
+    // Copy configuration file but replace the node ID with "server-id"
+    // Read the configuration file and replace the node ID with "server-id"
+    dora_cli::command::Command::Start(Start {
+        dataflow,
+        name: Some(node_id.to_string()),
+        coordinator_addr: IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
+        coordinator_port: DORA_COORDINATOR_PORT_CONTROL_DEFAULT,
```

Comment:
So this only works when the node is running on the same machine than the coordinator. This seems quite limiting.

Reply:
Agreed I will make this configurable.